Math/NumberTheory/SmoothNumbers.hs

```diff
@@ -130,24 +147,33 @@ smoothOverInRange s lo hi
 -- >>> import Data.Maybe
 -- >>> smoothOverInRangeBF (fromJust (fromList [2, 5])) 100 200
 -- [100, 125, 128, 160, 200]
-smoothOverInRangeBF :: forall a. Integral a => SmoothBasis a -> a -> a -> [a]
+smoothOverInRangeBF 
+  :: forall a. (Enum a, E.Euclidean a)
+  => SmoothBasis a
+  -> a
+  -> a
+  -> [a]
 smoothOverInRangeBF prs lo hi
   = coerce
-  $ filter (mf prs')
+  $ filter (isSmooth prs)
   $ coerce [lo..hi]
-  where
-    mf :: [WrappedIntegral a] -> WrappedIntegral a -> Bool
-    mf _         0 = False
-    mf []        n = n == 1 -- mf means manually factor
-    mf pl@(p:ps) n = if mod n p == 0
-                     then mf pl (div n p)
-                     else mf ps n
-    prs'           = coerce $ unSmoothBasis prs
 
 -- | isValid assumes that the list is sorted and unique and then checks if the list is suitable to be a SmoothBasis.
-isValid :: Euclidean a => [a] -> Bool
+isValid :: E.Euclidean a => [a] -> Bool
 isValid pl = length pl /= 0 && v' pl
   where
-    v' :: Euclidean a => [a] -> Bool
+    v' :: E.Euclidean a => [a] -> Bool
     v' []     = True
-    v' (x:xs) = x /= 0 && abs x /= 1 && abs x == x && all (coprime x) xs && v' xs
+    v' (x:xs) = x /= 0 && abs x /= 1 && abs x == x && all (E.coprime x) xs && v' xs
+
+-- | @isSmooth@ checks if a given number is smooth under a certain @SmoothBasis@.
+-- Does not check if the @SmoothBasis@ is valid.
+isSmooth :: forall a . E.Euclidean a => SmoothBasis a -> a -> Bool
+isSmooth prs x = mf (unSmoothBasis prs) x
+  where
+    mf :: [a] -> a -> Bool
+    mf _         0 = False
+    mf []        n = n == 1 -- mf means manually factor
```

Comment:
Let us relax this check to `abs n == 1`.

Reply:
@Bodigrim good catch. Negative numbers can be smooth.